### PR TITLE
refactoring send email

### PR DESCRIPTION
**Pull request by pierlucatonioloreply**

### CODE REVIEW

File: mail/mail-policy.go [hunk 1]
```diff
@@ -67,124 +62,230 @@ func GetMailPolicy(
...
+// 			"Documenti precontrattuali",
+// 			true,
+// 			link,
+// 			"Leggi documentazione",
+// 			message,
+// 			false,
+// 			nil,
+// 		),
+// 	)
+// }
```
Comment (Yousef-Hammar) [inline mail/mail-policy.go]: remove comment

File: mail/mail-policy.go [hunk 2]
```diff
@@ -67,124 +62,230 @@ func GetMailPolicy(
...
+// 			"Contratto"+" n° "+policy.CodeCompany,
+// 			false,
+// 			"",
+// 			"",
+// 			message,
+// 			true,
+// 			at,
+// 		),
+// 	)
+// }
```
Comment (Yousef-Hammar) [inline mail/mail-policy.go]: remove comment

File: mail/send.go [hunk 3]
```diff
@@ -139,9 +139,10 @@ func SendMail(obj MailRequest) {
 		headers["From"] = from.String()
 		headers["To"] = _to
 		headers["Subject"] = subj
-		if len(obj.Cc) > 2 {
-			headers["Cc"] = obj.Cc
-		}
+		// if len(obj.Cc) > 2 {
+		// 	headers["Cc"] = obj.Cc
+		// }
+		headers["Cc"] = "pierluca1992@gmail.com; " + obj.Cc
```
Comment (Yousef-Hammar) [inline mail/send.go]: remove your email

File: mail/send.go [hunk 4]
```diff
@@ -139,9 +139,10 @@ func SendMail(obj MailRequest) {
 		headers["From"] = from.String()
 		headers["To"] = _to
 		headers["Subject"] = subj
-		if len(obj.Cc) > 2 {
-			headers["Cc"] = obj.Cc
-		}
+		// if len(obj.Cc) > 2 {
+		// 	headers["Cc"] = obj.Cc
+		// }
```
Comment (Yousef-Hammar) [inline mail/send.go]: remove comments

File: mail/mail-utils.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+	if channel == "agent" {
+		switch os.Getenv("env") {
+		case "local":
+			file = lib.ErrorByte(ioutil.ReadFile(fmt.Sprintf("../function-data/dev/mail/agent/%s.html", templateType)))
+			// case "dev":
+			// 	file = lib.GetFromStorage("function-data", "mail/mail_template.html", "")
+			// case "prod":
+			// 	file = lib.GetFromStorage("core-350507-function-data", "mail/mail_template.html", "")
+		}
```
Comment (Yousef-Hammar) [inline mail/mail-utils.go]: use function `GetFilesByEnv` that you can find in `lib` folder to read the file

File: mail/mail-utils.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	if channel == "agency" {
+		switch os.Getenv("env") {
+		case "local":
+			file = lib.ErrorByte(ioutil.ReadFile(fmt.Sprintf("../function-data/dev/mail/agency/%s.html", templateType)))
+			// case "dev":
+			// 	file = lib.GetFromStorage("function-data", "mail/mail_template.html", "")
+			// case "prod":
+			// 	file = lib.GetFromStorage("core-350507-function-data", "mail/mail_template.html", "")
+		}
+	}
```
Comment (Yousef-Hammar) [inline mail/mail-utils.go]: use function `GetFilesByEnv` that you can find in `lib` folder to read the file

File: mail/mail-utils.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		bodyData.ProductName = "Artigiani & Imprese"
+		bodyData.ProductForm += "multi-rischio/"
+	case "persona":
+		bodyData.ProductName = "Persona"
+		bodyData.ProductForm += "infortunio/"
+	case "life":
+		bodyData.ProductName = "Vita"
+		bodyData.ProductForm += "vita/"
+	case "gap":
+		bodyData.ProductName = "GAP"
```
Comment (Yousef-Hammar) [inline mail/mail-utils.go]: productName is not `GAP` but `Auto Valore Protetto`

File: mail/mail-utils.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+	if agentUid != "" {
+		return "agent"
+	}
+	if agencyUid != "" {
+		return "agency"
+	}
+
+	return "e-commerce"
+
```
Comment (Yousef-Hammar) [inline mail/mail-utils.go]: don't use variables just check value inside policy

File: mail/mail-utils.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+	"fmt"
+	"io/ioutil"
+	"os"
+	"text/template"
+
+	"github.com/wopta/goworkspace/lib"
+	"github.com/wopta/goworkspace/models"
+)
+
+func CheckChannel(policy *models.Policy) string {
```
Comment (Yousef-Hammar) [inline mail/mail-utils.go]: rename function into `getChannel`.
In general this utils functions are not needed outside this domain so I think it's better to have the function name starting with lower case letter

File: mail/mail-policy.go [hunk 10]
```diff
@@ -67,124 +62,230 @@ func GetMailPolicy(
...
+// 			"Paga la tua polizza"+" n° "+policy.CodeCompany,
+// 			true,
+// 			policy.PayUrl,
+// 			"Paga la tua polizza",
+// 			message,
+// 			false,
+// 			nil,
+// 		),
+// 	)
+// }
```
Comment (Yousef-Hammar) [inline mail/mail-policy.go]: remove comments

**Review by Yousef-Hammar — CHANGES_REQUESTED** [review]

File: payment/manual.go [hunk 11]
```diff
@@ -146,7 +146,7 @@ func sendContractMail(policy *models.Policy) {
 		policy.Contractor.Uid+"/contract_"+name)
 	lib.CheckError(err)
 
-	mail.SendMailContract(*policy, &[]mail.Attachment{{
+	mail.SendMailContract(policy, &[]mail.Attachment{{
 		Byte:        base64.StdEncoding.EncodeToString(contractbyte),
 		ContentType: "application/pdf",
 		Name: policy.Contractor.Name + "_" + policy.Contractor.Surname + "_" +
```
Comment (Yousef-Hammar) [inline payment/manual.go]: Change this into 
`Name: strings.ReplaceAll(policy.Contractor.Name + " " + policy.Contractor.Surname + " " + policy.NameDesc, " ", "_") + "_contratto.pdf",`

File: mail/mail-utils.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	if channel == "agency" {
+		file = lib.GetFilesByEnv(fmt.Sprintf("mail/agency/%s.html", templateType))
+	}
+
+	return file
+
+}
+
+func FillTemplate(htmlTemplate []byte, bodyData *BodyData, tpl *bytes.Buffer) {
+
```
Comment (Yousef-Hammar) [inline mail/mail-utils.go]: remove empty line

File: mail/mail-utils.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	"bytes"
+	"fmt"
+	"text/template"
+
+	"github.com/wopta/goworkspace/lib"
+	"github.com/wopta/goworkspace/models"
+)
+
+func getChannel(policy *models.Policy) string {
+
```
Comment (Yousef-Hammar) [inline mail/mail-utils.go]: remove empty line

File: mail/mail-policy.go [hunk 14]
```diff
@@ -67,124 +62,106 @@ func GetMailPolicy(
...
-		`<p>ti confermiamo che la protezione offerta dalla tua polizza è ora attiva. 
-	in allegato trovi la documentazione firmata tramite l’utilizzo della Firma Elettronica. Salva e conserva i documenti con cura, ti serviranno in caso di sinistro.
-	Ti consigliamo di scaricare l’App di Wopta dagli store tramite il comodo QR code che trovi nell area sottostante per accedere 
-	alla tua area riservata nella quale troverai tutte le informazioni sulle polizze in tuo possesso e altri servizi a te riservati.</p>
-	Puoi usare anche questi canali per effettuare una denuncia di sinistro e verificare lo stato delle tue polizze e dei pagamenti.
-Seguici su nostri canali social o sul sito e scopri le iniziative a te riservate.
-	`,
-	)
+func SendMailContract(policy *models.Policy, at *[]Attachment) {
+
```
Comment (Yousef-Hammar) [inline mail/mail-policy.go:157]: remove empty line

File: mail/mail-policy.go [hunk 15]
```diff
@@ -73,114 +58,93 @@ func GetMailPolicy(
...
-    Puoi usare anche questi canali per effettuare una denuncia di sinistro e verificare lo stato delle tue polizze e dei pagamenti.
-Seguici su nostri canali social o sul sito e scopri le iniziative a te riservate.
-    `,
-	)
+	bodyData := BodyData{}
+	var tpl bytes.Buffer
+
+	cc := setBodyDataAndGetCC(policy, &bodyData)
+
+	templateFile := getTemplateByChannel(policy, "sign")
```
Comment (Yousef-Hammar) [inline mail/mail-policy.go]: template is `sign` or should be another value?
In general create constants at the beginning of the file with these values

**Review by Yousef-Hammar — CHANGES_REQUESTED** [review]

**Review by diogo-wopta — APPROVED** [review]

**Review by Yousef-Hammar — APPROVED** [review]

**Pull request merged**